### ISSUE TITLE
Can't override operations unless they came from a container

### ISSUE THREAD

**Issue by htp** [dry-rb/dry-transaction#75]
Using dry-transaction 0.10.2 (and Ruby 2.3.4p301, if it matters), I'm able to override operations only if they came from a container.

That is, this works:

```ruby
require 'dry/transaction'

Container = Dry::Container.new.tap do |c|
  c.register(:container_operation, call: false) do |_|
    Dry::Monads.Right(:success)
  end
end

class Example1
  include Dry::Transaction(container: Container)

  step :operation, with: :container_operation
end

puts Example1.new.call(nil)
# => Right(:success)

puts Example1.new(operation: ->(_) { Dry::Monads.Left(:error) }).call(nil)
# => Left(:error)
```

But this does not:

```ruby
require 'dry/transaction'

class Example2
  include Dry::Transaction

  step :operation

  def operation(_)
    Right(:success)
  end
end

puts Example2.new.call(nil)
# => Right(:success)

puts Example2.new(operation: ->(_) { Dry::Monads.Left(:error) }).call(nil)
# => Right(:success)
```

I expected the `Example2` case to behave the same way as the `Example1` case. Is this a known limitation or expected behavior of class-based transactions?

**Comment by GustavoCaso** [comment]
@htp I will have a look at it

**Comment by htp** [comment]
Let me know if there’s anything I can do to help.

**Comment by GustavoCaso** [comment]
@htp I have been investigating and looks like in the actual implementation it preferences the methods defined in the class itself over the ones injected.

In the first examples works because since is not defined in the class, but in the second example is defined in the class and also injected, It will pick the one from the class. https://github.com/dry-rb/dry-transaction/blob/master/lib/dry/transaction/instance_methods.rb#L16

Are you experience this behavior in your current app? How are you using dry-transaction?

**Comment by htp** [comment]
> I have been investigating and looks like in the actual implementation it preferences the methods defined in the class itself over the ones injected.

Thanks for taking a look!

> Are you experience this behavior in your current app?

Yes, I am.

> How are you using dry-transaction?

I wanted to inject some operations for testing purposes, exactly as documented on the [injecting operations](http://dry-rb.org/gems/dry-transaction/injecting-operations/) page in dry-transaction's documentation.

I'm currently using only class-based transactions, and I'm okay with working around the issue by defining operations in my container. But: for the sake of consistent (with non-class-based transactions) and non-surprising behavior, I think it would make more sense for injected operations to always take precedence, regardless of whether the transaction is class-based or non-class-based.

Is there a reason why they shouldn't? Maybe a design decision I'm not thinking of?

**Comment by htp** [comment]
> Maybe a design decision I'm not thinking of?

Ah- looks like [wrapping operations](http://dry-rb.org/gems/dry-transaction/wrapping-operations/) is why.

**Comment by GustavoCaso** [comment]
Indeed @htp.
Sorry for the late response

**Comment by GustavoCaso** [comment]
@htp should we close this issue?

**Comment by htp** [comment]
I feel like injected operations should take precedence over all other operations, regardless of whether the other operations were defined as instance methods or were resolved from a container.

I suppose that's more of a change in behavior than it is a bug, so yeah- I'll close the issue. If I'm able to find some free time, I'll send a PR your way :)

**Comment by GustavoCaso** [comment]
I would also have a look at it Thanks for the feedback @htp

**Comment by timriley** [comment]
Hi @htp, thanks for submitting this, and thanks also @GustavoCaso for talking this through!

It's an interesting behaviour you came across, @htp. I'd never thought about supporting replacement of instance method-based step operations by injected objects because I considered those instance methods to be "intrinsic" to the class, i.e. not separate enough to be some sort of external concern/external object. i.e. not the kind of thing that _should_ be changeable from the outside. Providing alternative objects for these when testing almost feels as much as a "smell" as e.g. stubbing private methods when testing.

So this is why we didn't support this behaviour when moving to the class-based transactions. You're right that this behaviour is different to how we handle container-based operations, but I wouldn't go so far as to say it's a regression (I know you weren't suggesting this, btw) — it's really the fact that instance method-based step operations are a wholly new thing and come with some slightly different behaviour.

---

So that's my feeling when looking at this from _one_ angle. But after considering this further, I feel like there's perhaps another angle — what if someone was using dry-transaction without any container at all? That person may want to define transactions consisting of steps using injected operation objects, and still have the ability to wrap those in local methods (or not).

For _this_ kind of usage, we really do need to support injected operation objects for steps defined without any associated container identifier (whether the step operation is provided by a local method or otherwise).

So, after thinking all of this through, I'd definitely welcome a PR to add this behaviour. What do you think?

I'm going to at least re-open this issue in the meantime.

**Comment by timriley** [comment]
However, I think if we did implement this behaviour, it wouldn't be able to satisfy your use case in your original issue description. Because I'd still want instance methods to behave in the same way for all steps — i.e. they're always called first, before any operation object, and a call to `super` from within the instance method will call the operation object.

So this won't allow an injected operation object to fully _replace_ a matching instance method. That method always will always get called first.

As a change, this does help dry-transaction work better without a container (i.e. such transactions would require all their step operation objects to be manually injected), which I think is a flexibility win.

So I'll leave it up to you as to whether you still want to go ahead and work on a PR yourself. Let me know either way :)

**Comment by htp** [comment]
> I'd still want instance methods to behave in the same way for all steps — i.e. they're always called first, before any operation object, and a call to super from within the instance method will call the operation object.

That sounds like a reasonable compromise :)

> So I'll leave it up to you as to whether you still want to go ahead and work on a PR yourself. Let me know either way :)

I'd like the chance to, please! Life is a bit hectic right now, so I won't be able to sit down and properly work on this for at least a week, but it *is* something I continue to think about, and I'd love to contribute to dry-rb.

**Comment by timriley** [comment]
@htp I'd love for you to contribute this, happy for you to get to it when you have the time :)

**Comment by htp** [comment]
(I'm still planning to work on this!)

**Comment by timriley** [comment]
@htp How you going with this one? Just checking in :)

**Comment by htp** [comment]
I swear I'm going to work on this, hah.

Been working toward / prepping for a major release at the office and that's taken priority- if you'd like to close this for triaging / hygiene, I can reopen when I actually have a change in hand for you.

**Comment by madis** [comment]
I'm interested in support for this functionality as well.

I stumbled on this behavioural difference as well when trying to refactor part of a system to use dry-transactions. I was going to do simple class based implementation first before deciding whether to extract individual steps to operations when I noticed it.

One of my goals is to be able to create a trace of transaction's execution (e.g. which steps succeed and where it fails for auditing - either by matching on the final transaction result or subscribing to individual steps. 

So when testing I'd like to modify my transaction class behaviour by injecting individual steps (i.e. having them fail/succeed) and checking that the trace got created as expected.

**Comment by htp** [comment]
(Not dead! I have a 20-hour flight coming up this weekend- I'll finally see about getting you a diff then.)

**Comment by timriley** [comment]
Thanks @htp, good luck!

**Comment by timriley** [comment]
Fixed in #109
